### PR TITLE
core/service: set up credentials for all start-post commands too

### PR DESCRIPTION
**Pull request by YHNdnzj**
Fixes https://github.com/systemd/systemd/issues/31194

### CODE REVIEW

File: src/core/service.c [hunk 1]
```diff
@@ -2466,7 +2468,7 @@ static void service_enter_condition(Service *s) {
                 r = service_spawn(s,
                                   s->control_command,
                                   s->timeout_start_usec,
-                                  EXEC_APPLY_SANDBOXING|EXEC_APPLY_CHROOT|EXEC_IS_CONTROL|EXEC_APPLY_TTY_STDIN,
+                                  EXEC_IS_CONTROL|EXEC_APPLY_TTY_STDIN,
```
Comment (yuwata) [inline src/core/service.c]: Here, EXEC_WRITE_CREDENTIALS is missing.

File: src/core/service.c [hunk 2]
```diff
@@ -2577,7 +2579,7 @@ static void service_enter_reload(Service *s) {
                 r = service_spawn(s,
                                   s->control_command,
                                   s->timeout_start_usec,
-                                  EXEC_APPLY_SANDBOXING|EXEC_APPLY_CHROOT|EXEC_IS_CONTROL|EXEC_CONTROL_CGROUP,
+                                  EXEC_IS_CONTROL|EXEC_CONTROL_CGROUP,
```
Comment (yuwata) [inline src/core/service.c]: Here also.

**Review by yuwata — CHANGES_REQUESTED** [review]
I think, we should introduce, something like the following.
```c
static ExecFlags service_exec_flags(ServiceState state, ServiceExecCommand command_id) {
        ...
}
```
Otherwise, we can easily introduce similar bug in the future.

File: src/core/service.c (hunk 1 above)
Comment (YHNdnzj) [reply src/core/service.c]: We don't want `ExecCondition=` to access credentials IIRC. Generally it should be pretty irrelevant from other service processes.

File: src/core/service.c (hunk 2 above)
Comment (YHNdnzj) [reply src/core/service.c]: ExecReload= should signal the main process instead of accessing things like credentials too.

**Comment by YHNdnzj** [comment]
I'll introduce a helper like that. But I think ExecCondition= and -Reload= by design should not be accessing credentials.

File: src/core/service.c (hunk 1 above)
Comment (yuwata) [reply src/core/service.c]: Then, I may misunderstand the flags in `service_run_next_control()`.

File: src/core/service.c (hunk 2 above)
Comment (YHNdnzj) [reply src/core/service.c]: Uh, we set that in `service_run_next_control`...

File: src/core/execute.c [hunk 3]
```diff
@@ -309,13 +308,14 @@ int exec_params_get_cgroup_path(
          * this is not necessary, the cgroup is still empty. We distinguish these cases with the EXEC_CONTROL_CGROUP
          * flag, which is only passed for the former statements, not for the latter. */
 
-        if (FLAGS_SET(params->flags, EXEC_CGROUP_DELEGATE) && (FLAGS_SET(params->flags, EXEC_CONTROL_CGROUP) || c->delegate_subgroup)) {
+        if (FLAGS_SET(params->flags, EXEC_CGROUP_DELEGATE|EXEC_CONTROL_CGROUP) || c->delegate_subgroup) {
```
Comment (yuwata) [inline src/core/execute.c]: This logically changes the behavior. Is it intended?

File: src/core/service.c [hunk 4]
```diff
@@ -2631,14 +2665,9 @@ static void service_run_next_control(Service *s) {
         pidref_done(&s->control_pid);
 
         r = service_spawn(s,
+                          s->control_command_id,
                           s->control_command,
                           timeout,
-                          EXEC_APPLY_SANDBOXING|EXEC_APPLY_CHROOT|EXEC_IS_CONTROL|
-                          (IN_SET(s->state, SERVICE_CONDITION, SERVICE_START_PRE, SERVICE_START, SERVICE_START_POST, SERVICE_RUNNING, SERVICE_RELOAD) ? EXEC_WRITE_CREDENTIALS : 0)|
```
Comment (yuwata) [inline src/core/service.c:2637]: Here, EXEC_WRITE_CREDENTIALS is set when the state is `SERVICE_CONDITION`. So, I think we should enable credentials for commands `SERVICE_EXEC_CONDITION`, no?

File: src/core/service.c [hunk 5]
```diff
@@ -1594,15 +1595,51 @@ static Service *service_get_triggering_service(Service *s) {
...
+        assert(command_id >= 0);
+        assert(command_id < _SERVICE_EXEC_COMMAND_MAX);
+
+        /* Control processes spawned before main process also get tty access */
+        if (IN_SET(command_id, SERVICE_EXEC_CONDITION, SERVICE_EXEC_START_PRE, SERVICE_EXEC_START, /* FIXME */ SERVICE_EXEC_STOP_POST))
+                flags |= EXEC_APPLY_TTY_STDIN;
+
+        /* All start phases get access to credentials */
+        // FIXME: SERVICE_EXEC_START_POST
+        if (IN_SET(command_id, SERVICE_EXEC_START_PRE, SERVICE_EXEC_START, /* FIXME */ SERVICE_RELOAD))
```
Comment (yuwata) [inline src/core/service.c]: `SERVICE_RELOAD` -> `SERVICE_EXEC_RELOAD`, though this will be dropped in the later commit.

**Review by yuwata — CHANGES_REQUESTED** [review]

File: src/core/service.c (hunk 4 above)
Comment (YHNdnzj) [reply src/core/service.c:2637]: As mentioned in commit msg, I think this is accidentally introduced in https://github.com/systemd/systemd/pull/27279. I don't think this is desirable, and the first `ExecCondition=` cannot access creds anyway. So the incompatibility introduced should be negligible.

File: src/core/execute.c (hunk 3 above)
Comment (YHNdnzj) [reply src/core/execute.c]: ah, right...

File: src/core/service.c (hunk 4 above)
Comment (yuwata) [reply src/core/service.c:2637]: OK, but the commit message mention ExecReload. Please also mention ExecCondition somewhere.

Comment (yuwata) [reply src/core/service.c:2637]: Oh, in the revised commit, it is mentioned. Thank you.

**Review by yuwata — APPROVED** [review]

**Comment by YHNdnzj** [comment]
(Pushed to update commit messages)

**Pull request merged**